src/vs/workbench/contrib/terminal/browser/terminalView.ts

```diff
@@ -477,3 +506,45 @@ function getSingleTabTooltip(instance: ITerminalInstance | null): string {
 	}
 	return `${instance.title} ${instance.shellLaunchConfig.description}`;
 }
+
+class TerminalThemeIconStyle extends Themable {
+	private _styleElement: HTMLElement;
+	constructor(
+		container: HTMLElement,
+		@IThemeService private readonly _themeService: IThemeService,
+		@ITerminalService private readonly _terminalService: ITerminalService
+	) {
+		super(_themeService);
+		this._registerListeners();
+		this._styleElement = document.createElement('style');
+		container.appendChild(this._styleElement);
+		this._register(toDisposable(() => container.removeChild(this._styleElement)));
+		this.updateStyles();
+	}
+
+	private _registerListeners(): void {
+		this._register(this._terminalService.onInstanceIconChanged(() => this.updateStyles()));
+	}
+
+	override updateStyles(): void {
+		super.updateStyles();
+		let css = '';
+		for (const instance of this._terminalService.terminalInstances) {
+			const icon = instance.icon;
+			if (!icon) {
+				return;
+			}
+			const uri = icon instanceof URI ? icon :
+				icon instanceof Object && 'light' in icon && 'dark' in icon ?
+					(this._themeService.getColorTheme().type === ColorScheme.LIGHT ?
+						icon.light
+						: icon.dark) : undefined;
+			if (uri instanceof URI) {
+				let uriIconKey = hash(uri.path).toString(36);
```

Comment:
Getting the hash from terminalIcon.ts will make sure they don't accidentally differ